libs/@blockprotocol/graph/src/stdlib/subgraph/element/map-revisions.ts

```diff
@@ -0,0 +1,43 @@
+import { Vertex } from "../../../types/subgraph/vertices.js";
+
+type BaseIdToRevisions<
+  TemporalSupport extends boolean,
+  GraphElementType extends Vertex<TemporalSupport>["inner"],
+> = Record<
+  /*
+   * @todo - we _should_ be able to use `Extract<GraphElementForIdentifier<TemporalSupport, VertexId<any, any>>`
+   *   here to actually get a strong type (like `EntityId` or `BaseUri`). TypeScript seems to break on using it with a
+   *   generic though. So for now we write `string` because all of the baseId's of `VertexId` are string aliases anyway.
+   */
+  string,
+  GraphElementType[]
+>;
+
+/**
+ * Takes a collection of graph elements, and returns an object that groups them by their base IDs, mapping the IDs to
+ * the collection of revisions.
+ *
+ * @param elements
+ */
+export const mapElementsIntoRevisions = <
+  TemporalSupport extends boolean,
+  GraphElementType extends Vertex<TemporalSupport>["inner"],
+>(
+  elements: GraphElementType[],
+): BaseIdToRevisions<TemporalSupport, GraphElementType> => {
+  return elements.reduce((revisionMap, element) => {
+    const baseId =
+      "entityId" in element.metadata.recordId
+        ? element.metadata.recordId.entityId
+        : element.metadata.recordId.baseUri;
+
+    if (revisionMap[baseId] !== undefined) {
+      revisionMap[baseId]!.push(element);
+    } else {
+      // eslint-disable-next-line no-param-reassign
+      revisionMap[baseId] = [element];
+    }
```

Comment:
` ??= ` I had not seen that operator before, interesting.

I appreciate this is about as micro an optimisation as you can possibly get, but it seemed that it would be faster to create an array with the thing in it rather than allocate an array and then push the thing in. But I've done some more reading on JS arrays and the various implementations across the JS engines and my brain has broken, so I'm not even sure the intuition was correct. (sparse arrays, optimisations for numeric arrays, not actually allocating contiguous space despite setting initial length to 1000, etc.)

Regardless, absolutely wouldn't argue for the previous code on the basis of that tiny a potential optimisation.

[Improve non-existing array initialization](https://github.com/blockprotocol/blockprotocol/pull/928/commits/262fddead22c34c0268ce273dfdeac72bad535f8)

Reply:
It's 'new' as of 2020. There are also `||=` and `&&=` but I've never had occasion to use those.